### PR TITLE
SECURITY: Default tags to show count of topics in unrestricted categories

### PR DESCRIPTION
**Pull request by tgxworld**
## What is the problem here? 

Currently, `Tag#topic_count` is a count of all regular topics regardless of whether the topic is in a read restricted category or not. As a result, any users can technically poll a sensitive tag to determine if a new topic is created in a category which the user has not excess to. We classify this as a minor leak in sensitive information.

## What does this PR change? 

1. Introduce `Tag#public_topic_count` which only count topics which have been tagged with a given tag in public categories.
2. Rename `Tag#topic_count` to `Tag#staff_topic_count` which counts the same way as `Tag#topic_count`. In other words, it counts all topics tagged with a given tag regardless of the category the topic is in. The rename is also done so that we indicate that this column contains sensitive information. 
3. Change all previous spots which relied on `Topic#topic_count` to rely on `Tag.topic_column_count(guardian)` which will return the right "topic count" column to use based on the current scope. 
4. Introduce `SiteSetting.include_secure_categories_in_tag_counts` site setting to allow site administrators to always display the tag topics count using `Tag#staff_topic_count` instead.

## Notes for reviewer
1. It is easier to follow this PR if you follow the commits in the order they were made.
1. This was discussed internally and our objective here is not to display the count as accurately as possible. Making the count accurate for all users is complicated to do and we're not ready to make that investment at this point in time.

### CODE REVIEW

File: db/migrate/20230118020114_add_public_topic_count_to_tags.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      FROM tags
+      INNER JOIN topic_tags ON tags.id = topic_tags.tag_id
+      INNER JOIN topics ON topics.id = topic_tags.topic_id AND topics.deleted_at IS NULL AND topics.archetype != 'private_message'
+      INNER JOIN categories ON categories.id = topics.category_id AND NOT categories.read_restricted
+      GROUP BY tags.id
+    ) x
+    WHERE x.tag_id = t.id
+    AND x.topic_count <> t.public_topic_count;
+    SQL
+  end
```
Comment (tgxworld) [inline db/migrate/20230118020114_add_public_topic_count_to_tags.rb:23]: Note that this is just copying the SQL from `Tag.ensure_consistency!` so I didn't bother with a migration test as `Tag.ensure_consistency!` is well tested.

File: db/migrate/20230119000943_add_staff_topic_count_to_tags.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      LEFT JOIN topic_tags ON tags.id = topic_tags.tag_id
+      LEFT JOIN topics ON topics.id = topic_tags.topic_id
+      AND topics.deleted_at IS NULL
+      AND topics.archetype != 'private_message'
+      GROUP BY tags.id
+    ) x
+    WHERE x.tag_id = t.id
+    AND x.topic_count <> t.staff_topic_count
+    SQL
+  end
```
Comment (tgxworld) [inline db/migrate/20230119000943_add_staff_topic_count_to_tags.rb:22]: Note that this is just copying the SQL from `Tag.ensure_consistency!` so I didn't bother with a migration test as this is already tested as part of `Tag.ensure_consistency!`.

File: db/post_migrate/20230119024157_remove_topic_count_from_tags.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require "migration/column_dropper"
+
+class RemoveTopicCountFromTags < ActiveRecord::Migration[7.0]
+  DROPPED_COLUMNS ||= { tags: %i[topic_count] }
+
+  def up
+    DROPPED_COLUMNS.each { |table, columns| Migration::ColumnDropper.execute_drop(table, columns) }
```
Comment (tgxworld) [inline db/post_migrate/20230119024157_remove_topic_count_from_tags.rb:9]: Usually we like to leave columns around but this column and its results can easily be rebuilt so there is no need for us to play it safe here.

File: spec/integration/tag_counts_spec.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+RSpec.describe "Updating tag counts" do
```
Comment (tgxworld) [inline spec/integration/tag_counts_spec.rb:3]: Integration test for making sense of what happens when regular topics are moved into different categories or when they are converted to private messages.

File: db/post_migrate/20230119024157_remove_topic_count_from_tags.rb (hunk 3 above)
Comment (martin-brennan) [reply db/post_migrate/20230119024157_remove_topic_count_from_tags.rb:9]: Yes I agree, it's not really worthwhile information that can't be recreated, thanks for adding the note :+1:

File: spec/integration/tag_counts_spec.rb (hunk 4 above)
Comment (martin-brennan) [reply spec/integration/tag_counts_spec.rb:3]: Hah I had to add one of these for my secure upload change to deal with the more complicated scenarios too :+1:

File: config/locales/server.en.yml [hunk 5]
```diff
@@ -1651,6 +1651,7 @@ en:
     content_security_policy_frame_ancestors: "Restrict who can embed this site in iframes via CSP. Control allowed hosts on <a href='%{base_path}/admin/customize/embedding'>Embedding</a>"
     content_security_policy_script_src: "Additional allowlisted script sources. The current host and CDN are included by default. See <a href='https://meta.discourse.org/t/mitigate-xss-attacks-with-content-security-policy/104243' target='_blank'>Mitigate XSS Attacks with Content Security Policy.</a>"
     invalidate_inactive_admin_email_after_days: "Admin accounts that have not visited the site in this number of days will need to re-validate their email address before logging in. Set to 0 to disable."
+    include_secure_categories_in_tag_counts: "When enabled, count of topics tagged with a given tag will include topics that are in read restricted categories."
```
Comment (martin-brennan) [inline config/locales/server.en.yml:1654]: Could add a clarifying comment about the opposite. "When enabled, only staff will have these topics included in tag counts."

**Review by martin-brennan — APPROVED** [review]
Looks great, I mostly had questions and minor nitpicks so will let you clean them up then merge :+1:

File: app/models/tag.rb [hunk 6]
```diff
@@ -76,6 +76,25 @@ def self.update_topic_counts
          AND x.topic_count <> t.topic_count
     SQL
 
+    DB.exec <<~SQL
+      UPDATE tags SET public_topic_count = 0 WHERE tags.public_topic_count <> 0;
```
[the remark this answers is not in the document]
Comment (tgxworld) [reply app/models/tag.rb:87]: The query that sets the count below excludes tags that are not used in any public topics. It was too complicated to get the query to do everything so I split it up into two queries instead. Hmm but now that you mention it, it is kind of inefficient to have to reset the count when we don't need to :thinking: Let me have a look at the query again.

File: app/models/topic_tag.rb [hunk 7]
```diff
@@ -27,12 +31,17 @@ class TopicTag < ActiveRecord::Base
...
         tag.decrement!(:pm_topic_count)
       else
-        if topic.category_id &&
-             stat = CategoryTagStat.find_by(tag_id: tag_id, category: topic.category_id)
-          stat.topic_count == 1 ? stat.destroy : stat.decrement!(:topic_count)
+        if stat = CategoryTagStat.find_by(tag_id: tag_id, category: topic.category_id)
+          stat.topic_count == 1 ? stat.destroy! : stat.decrement!(:topic_count)
+        end
+
+        counters_to_update = { topic_count: -1 }
```
[the remark this answers is not in the document]
Comment (tgxworld) [reply app/models/topic_tag.rb]: I think you might have been looking at the commit before the column was ignored. This was resolved in a latter commit :grin:

File: app/models/topic_tag.rb [hunk 8]
```diff
@@ -9,14 +9,18 @@ class TopicTag < ActiveRecord::Base
...
       else
-        tag.increment!(:topic_count)
-
-        if topic.category_id
-          if stat = CategoryTagStat.find_by(tag_id: tag_id, category_id: topic.category_id)
-            stat.increment!(:topic_count)
-          else
-            CategoryTagStat.create(tag_id: tag_id, category_id: topic.category_id, topic_count: 1)
-          end
+        counters_to_update = { topic_count: 1 }
```
[the remark this answers is not in the document]
Comment (tgxworld) [reply app/models/topic_tag.rb]: I think you might have been looking at the commit before the column was ignored. This was resolved in a latter commit :grin:

File: app/models/topic_tag.rb [hunk 9]
```diff
@@ -9,14 +9,18 @@ class TopicTag < ActiveRecord::Base
...
+
+        if Category.exists?(id: topic.category_id, read_restricted: false)
+          counters_to_update[:public_topic_count] = 1
+        end
+
+        Tag.update_counters(tag.id, counters_to_update)
+
+        if stat = CategoryTagStat.find_by(tag_id: tag_id, category_id: topic.category_id)
```
[the remark this answers is not in the document]
Comment (tgxworld) [reply app/models/topic_tag.rb:20]: `CategoryTagStat` is being used to determine the "top tags" for a site but it is currently only being used in the backend.

File: app/models/tag.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (tgxworld) [reply app/models/tag.rb:87]: Updated the query :+1:

File: app/models/topic.rb [hunk 10]
```diff
@@ -971,6 +972,15 @@ def changed_to_category(new_category)
 
         if old_category
           Category.where(id: old_category.id).update_all("topic_count = topic_count - 1")
+
+          count =
+            if old_category.read_restricted && !new_category.read_restricted
+              1
+            elsif !old_category.read_restricted && new_category.read_restricted
```
[the remark this answers is not in the document]
Comment (tgxworld) [reply app/models/topic.rb:979]: Yup those should be 0 which just means it is a noop.

File: app/models/tag.rb [hunk 11]
```diff
@@ -4,6 +4,10 @@ class Tag < ActiveRecord::Base
   include Searchable
   include HasDestroyedWebHook
 
+  self.ignored_columns = [
+    "topic_count", # TODO: Drop on 1 July 2023
```
[the remark this answers is not in the document]
Comment (tgxworld) [reply app/models/tag.rb:8]: Good point, have updated the PR with the changes requested here.

File: config/locales/server.en.yml (hunk 5 above)
Comment (tgxworld) [reply config/locales/server.en.yml:1654]: Done :+1:

**Pull request merged**